README.md

<div data-align="center">

<p align="center">

<img src="inst/app/www/iphone_mobilequake.png"  align=center>

</p>

<!-- README.md is generated from README.Rmd. Please edit that file -->

## mobilequake

<!-- badges: start -->

<!-- badges: end -->

Dashboard to track Earthquakes in real-time, based on [USGS
data](https://earthquake.usgs.gov/). The app is built with
[shinyMobile](https://rinterface.github.io/shinyMobile/) &
[echarts4r](https://echarts4r.john-coene.com/) &
[leaflet](http://rstudio.github.io/leaflet)

</div>

## DEMO

[Tracker](http://tools.dataatomic.com/shiny/mobilequake)

## Get it

You can download the package and run it locally or deploy it.

``` r
# install.packages("remotes")
remotes::install_github("korur/mobilequake")
```

## How to run

``` r
## Run the app
mobilequake::run_app()
```

<!--html_preserve-->

<div class="muted well" style="width: 100% ; height: 400px ; text-align: center; box-sizing: border-box; -moz-box-sizing: border-box; -webkit-box-sizing: border-box;">

Shiny applications not supported in static R Markdown documents

</div>

<!--/html_preserve-->


NEWS.md
# mobilequake 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(mobilequake)

test_check("mobilequake")


R/mod_map_module.R
# Module UI

#' @title   mod_map_module_ui and mod_map_module_server
#' @description  A shiny Module.
#'
#' @param id shiny id
#' @param input internal
#' @param output internal
#' @param session internal
#'
#' @rdname mod_map_module
#'
#' @keywords internal
#' @export 
#' @importFrom shiny NS tagList 
#' @importFrom stats time
mod_map_module_ui <- function(id){
  
  ns <- shiny::NS(id)
  
  f7Card(
    f7Slider(
      inputId = ns("hours"),
      label = HTML("<h5>","Show Earthquakes within selected hours","</h5>"),
      max = 168,
      min = 0,
      value = 72,
      scaleSteps = 7,
      scaleSubSteps = 1,
      scale = TRUE,
      color = "orange",
      labels = tagList(
        shinyMobile::f7Icon("circle", old=FALSE),
        shinyMobile::f7Icon("circle_fill", old=FALSE)
      )
    ))
}

# Module Server

#' @rdname mod_map_module
#' @export
#' @keywords internal

mod_map_module_server <- function(input, output, session){
  
  ns <- session$ns
  
  map <- reactive({
    url <- ("https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_week.csv")
    df <- readr::read_csv(url, col_types = readr::cols())
    dt <- as.POSIXlt(Sys.time(), tz = "UTC")
    # If no file is selected, don't do anything
    new.dt <- dt - as.difftime(input$hours, units="hours")
    df <- df %>% dplyr::filter(time >= new.dt)
    
    df$size <- cut(df$mag,breaks = c(-Inf, 3.9, 4.9, 5.9, 6.9, 7.9, Inf),
                   labels=c("minor", "light", "moderate", "strong", "major", "great 8+"))
    
    # Create popup in HTML
    pop <- paste("<h3>",
                 "<b>Place:</b>", df$place, "<br>",
                 "<b>Time:</b>", df$time, "<br>",   # Format nicer
                 "<b>Mag:</b>", as.character(df$mag), "<br>",
                 "<b>Depth:</b>", as.character(df$depth), "km<br>", "</h3>"
    )
    
    # Create colour pallet
    col_rainbow<- c("#66ffff","#1aff1a","#f07900","#ff0000","#b30000","#b30059")
    pallet <- leaflet::colorFactor(col_rainbow, df$size)
    
    map <- leaflet::leaflet(df, options = leaflet::leafletOptions(minZoom = 2)) %>% 
      
      leaflet::addProviderTiles(leaflet::providers$CartoDB.DarkMatter) %>%
      leaflet::setView(100.65, 120.0285, zoom = 1) %>%
      leaflet::setView(24, 10, zoom=2) %>%
      leaflet::addCircles( ~longitude, ~latitude,  
                  weight= ~ifelse(mag < 4, 1, 6),
                  color= ~pallet(size),
                  radius = ~ifelse(mag < 4, 2, 5), # add ifs
                  popup = pop
      )    %>% 
      leaflet::addLegend( "bottomright", pal = pallet,
                 values = sort(df$size),
                 title = "Magnitude")
  })
  return(map)
  
}

## To be copied in the UI
# mod_map_module_ui("map_module_ui_1")

## To be copied in the server
# callModule(mod_map_module_server, "map_module_ui_1")



R/mod_input_module.R
# Module UI
  
#' @title   mod_input_module_ui and mod_input_module_server
#' @description  A shiny Module.
#'
#' @param id shiny id
#' @param input internal
#' @param output internal
#' @param session internal
#'
#' @rdname mod_input_module
#'
#' @keywords internal
#' @export 
#' @importFrom shiny NS tagList 
mod_input_module_ui <- function(id){
  ns <- NS(id)
  tagList(
    f7Slider(
      inputId = ns("hours"),
      label = HTML("<h5>","Show Earthquakes within selected hours","</h5>"),
      max = 168,
      min = 0,
      value = 72,
      scaleSteps = 7,
      scaleSubSteps = 1,
      scale = TRUE,
      color = "orange",
      labels = tagList(
        shinyMobile::f7Icon("circle", old=FALSE),
        shinyMobile::f7Icon("circle_fill", old=FALSE)
      )
    )
  )
}
    
# Module Server
    
#' @rdname mod_input_module
#' @export
#' @keywords internal
    
mod_input_module_server <- function(input, output, session){
  ns <- session$ns
  return(
      hours = reactive({ input$hours })
    )
}
    
## To be copied in the UI
# mod_input_module_ui("input_module_ui_1")
    
## To be copied in the server
# callModule(mod_input_module_server, "input_module_ui_1")
 


R/mod_plots_module.R
# Module UI
  
#' @title   mod_plots_module_ui and mod_plots_module_server
#' @description  A shiny Module.
#'
#' @param id shiny id
#' @param input internal
#' @param output internal
#' @param session internal
#'
#' @rdname mod_plots_module
#'
#' @keywords internal
#' @export 
#' @importFrom shiny NS tagList 
mod_plots_module_ui <- function(id){
  ns <- NS(id)
  
  ns <- shiny::NS(id)
  tagList(
    
    f7Card(f7Col(leafletOutput(ns("mapp")))), f7Card(f7Col(echarts4rOutput(ns("timeline"))))
    
  )
}
    
# Module Server
    
#' @rdname mod_plots_module
#' @export
#' @keywords internal
    
mod_plots_module_server <- function(input, output, session, dataset, time_window){
  
  ns <- session$ns

  qqq <- reactive({
  selected_dt <- current_time - as.difftime(time_window(), units="hours")
  quake_selected <- quake_df %>% dplyr::filter(time >= selected_dt)
  return(quake_selected)
  })
  
  mapp <- reactive({
    
    # Create popup in HTML
    pop <- paste("<h3>",
                 "<b>Place:</b>", qqq()$place, "<br>",
                 "<b>Time:</b>", qqq()$time, "<br>",   # Format nicer
                 "<b>Mag:</b>", as.character(qqq()$mag), "<br>",
                 "<b>Depth:</b>", as.character(qqq()$depth), "km<br>", "</h3>"
    )
    
    # Create colour pallet
    col_rainbow<- c("#66ffff","#1aff1a","#f07900","#ff0000","#b30000","#b30059")
    pallet <- leaflet::colorFactor(col_rainbow, qqq()$size)
    
    # Create the map
    map <- leaflet::leaflet(qqq(), options = leaflet::leafletOptions(minZoom = 1)) %>% 
      
      leaflet::addProviderTiles(leaflet::providers$CartoDB.DarkMatter) %>%
      leaflet::addCircles( ~longitude, ~latitude,  
                           weight= ~ifelse(mag < 4, 1, 6),
                           color= ~pallet(size),
                           radius = ~ifelse(mag < 4, 2, 5), # add ifs
                           popup = pop) %>% 
      leaflet::addLegend( "bottomright", pal = pallet,
                          values = sort(qqq()$size)) %>% 
      leaflet::setView(lng = 70, lat = 15, zoom = 1) %>%
      leaflet::setMaxBounds( lng1 = -180 + .05, 
                             lat1 = -90 + .05, 
                             lng2 = 180 - .05, 
                             lat2 = 90 - .05) 
    return(map)
  })
  # Leaflet map output
  
  output$mapp <- leaflet::renderLeaflet({
    mapp()
  })
  
  # Earthquake timeline echarts4r output
  
  output$timeline <- renderEcharts4r({

   qqq() %>%  
      e_charts(time) %>%
      e_bar(mag) %>% 
      e_axis_labels(x = "Date",y= "Magnitute") %>% # axis labels
      e_title("Timeline" ) %>%  # Add title 
      e_theme("halloween") %>%  # theme
      e_legend(show = FALSE) %>%  # move legend to the bottom
      e_tooltip(trigger = "axis")  %>% #tooltip
      e_visual_map(mag, orient = "horizontal",
                   right = "center",
                   top = 5,
                   textStyle = list(color = "#fff"), scale = NULL)  
    
  }) 
  
  
}
    
## To be copied in the UI
# mod_plots_module_ui("plots_module_ui_1")
    
## To be copied in the server
# callModule(mod_plots_module_server, "plots_module_ui_1")
 


R/app_ui.R
#' @import shiny
#' @import shinyMobile
#' @import leaflet
#' @import echarts4r
#' @import shinyscroll
#' @import waiter
#' @import dplyr



url <- ("https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/all_week.csv")
quake_df <- readr::read_csv(url, col_types = readr::cols())
quake_df <- quake_df %>% filter(!is.na(quake_df$mag))
quake_df$size <- cut(quake_df$mag,breaks = c(-Inf, 3.9, 4.9, 5.9, 6.9, 7.9, Inf),
               labels=c("minor", "light", "moderate", "strong", "major", "great 8+"))
current_time <- as.POSIXlt(Sys.time(), tz = "UTC")

loader <- tagList(
  waiter::spin_3circles(),
  br(),br(),
  h3("Accessing real-time data...")
)

app_ui <- function() {
  version <- paste0("v", packageVersion("mobilequake"))
  
  tagList(
    # Leave this function for adding external resources
    golem_add_external_resources(),
    # List the first level UI elements here 
    f7Page(
      shinyscroll::use_shinyscroll(),
      sever::use_sever(),
      waiter::use_waiter(), # dependencies
      waiter::waiter_show_on_load(loader, color = "#000000"),
      title = "Earthquake Tracker",
      dark_mode = TRUE,
      init = f7Init(
        skin = "auto", 
        theme = "dark"
      ),
      f7SingleLayout(
        navbar = f7Navbar(
          title = "Earthquake Tracker",
          hairline = TRUE,
          shadow = TRUE,
          left_panel = TRUE,
          right_panel = FALSE
        ),
        f7Panel(
          title = "About", 
          side = "left", 
          theme = "dark",
          effect = "cover",
          p("Dashboard to track Earthquakes in real-time, based on USGS data. Built with shinyMobile & echarts4r & leaflet & Golem. The code is open-source so you can deploy it yourself."),
          f7Link(label = "Author", src = "https://dataatomic.com", external = TRUE),
          f7Link(label = "Code", src = "https://github.com/korur/mobilequake", external = TRUE),
          f7Link(label = "USGS", src = "https://earthquake.usgs.gov/", external = TRUE),
          f7Link(label = "Coronavirus Tracker", src = "http://tools.dataatomic.com/shiny/CoronaOutbreak/", external = TRUE)
        ),
        
        f7Card(mod_input_module_ui("input_module_ui_1")),
        mod_plots_module_ui("plots_module_ui_1")
      ))
    
  ) #taglist
}

#' @import shiny
golem_add_external_resources <- function(){
  
  addResourcePath(
    'www', system.file('app/www', package = 'mobilequake')
  )
  
  tags$head(
    golem::activate_js(),
    golem::favicon()
    # Add here all the external resources
    # If you have a custom.css in the inst/app/www
    # Or for example, you can add shinyalert::useShinyalert() here
    #tags$link(rel="stylesheet", type="text/css", href="www/custom.css")
  )
}
